mat-mat mult. with shared memory

artv3 opened the pull request:
This PR adds a variation of matrix-matrix multiplication using RAJA Teams and shared memory. 

Running the kernel variants gave me this (nvcc10-gcc7.3.1) 
```
Speedup Report (T_ref/T_var): ref var = Base_Seq  ,  ,  ,  ,  ,  ,  ,  ,  ,  ,                                                                                                                
Kernel                , Base_Seq , Lambda_Seq , RAJA_Seq , Base_OpenMP , Lambda_OpenMP , RAJA_OpenMP , Base_CUDA , Lambda_CUDA , RAJA_CUDA , RAJA_WORKGROUP_CUDA                              
Basic_MAT_MAT_SHARED  ,    1.000 ,      0.955 ,    0.945 ,      21.366 ,        18.798 ,      17.682 ,   327.116 ,     316.699 ,   348.101 ,             Not run    
```

artv3 commented:
Folks, there a couple of features of Teams::OpenMP I would like to use however; it requires updating RAJA. Shall I hold off on that? or Do we want update RAJA?

MrBurmark commented on src/basic/MAT_MAT_SHARED-Cuda.cpp:
```diff
@@ -0,0 +1,228 @@
+          block_x_direct(0, Nx, [&](Index_type bx) {
+            MAT_MAT_SHARED_BODY_0
+
+            thread_y_direct(0, TL_SZ, [&](Index_type ty) {
+              thread_x_direct(0, TL_SZ, [&](Index_type tx) { MAT_MAT_SHARED_BODY_1 });
+            });
```
I'm not sure that using functions to do the mapping here quite fits with the way things are done in the rest of the perf suite. Shouldn't this be the same as the direct cuda version with just the addition of lambdas?
```suggestion
            Index_type ty = threadIdx.y;
            Index_type tx = threadIdx.x;
            auto outer_body_1 = [&](Index_type ty) {
               auto inner_body_1 = [&](Index_type tx) { MAT_MAT_SHARED_BODY_1 };
               if (tx < TL_SZ) inner_body_1(tx);
            };
            if (ty < TL_SZ) outer_body_1(ty);
```

artv3 replied:
I was thinking it would be easier to read, but that's a good point. I'll switch to this style

artv3 commented:
Once https://github.com/LLNL/RAJAPerf/pull/120 is in, I want to make a small adjustment to the OMP variant.

artv3 commented:
Hi folks, just to follow up here. Is there anything else I can do?

MrBurmark commented on src/basic/MAT_MAT_SHARED-Cuda.cpp:
```diff
@@ -0,0 +1,260 @@
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+
+      mat_mat_shared<<<griddim, blockdim>>>(N, A, B, C);
```
error checking

MrBurmark commented on src/basic/MAT_MAT_SHARED-OMPTarget.cpp:
```diff
@@ -0,0 +1,44 @@
+
+  if (vid == Base_OpenMPTarget) {
+
+    startTimer();
+    for (RepIndex_type irep = 0; irep < run_reps; ++irep) {
+    }
+    stopTimer();
```
Should this just be empty, or even not here at all?

rhornung67 replied:
Is there no way to use device shared memory in OpenMP target offload?

artv3 replied:
Chatting with @rchen20 it might be a feature in OpenMP 5.1. It would be really cool if we can get that going. @trws what do you think? Has anyone tried it here?

trws replied:
It's technically supported farther back than that, 5.0 IIRC.  There is some argument about which forms are meant to work and which aren't, which is not great, but the one that's agreed to work by basically everyone (and I think is implemented in at least clang) looks like this:

```c
double my_shared_mem[size];
#pragma omp target teams distribute allocate(omp_cgroup_mem_alloc: my_shared_mem) private(my_shared_mem)
#pragma omp parallel for
<loop> // my_shared_mem is private to the team, separate for each contention group, so it should be shared/local memory
```

Constant memory should also be possible like this, but I have yet to test it anywhere:

```c
double my_const_mem[size];
#pragma omp target teams distribute allocate(omp_const_mem_alloc: my_shared_mem) firstprivate(my_const_mem)
#pragma omp parallel for
<loop> // my_shared_mem is private to the team, separate for each contention group, so it should be shared/local memory
```

Note this one has firstprivate, that's the only way to write into a constant memory segment allocated by that openmp memory allocator right now.

trws replied:
I'd note that both of the above, as well as the example below, are required to give you the *semantics* of shared memory on a gpu on any device where they execute.  Whether they actually use shared memory on a given GPU is QoI, so having a benchmark sitting around we can use to show when appropriate memories aren't being used might be really good.

Just using private at the teams level, or declaring a variable at that level, should also work.

artv3 replied:
@trws in regards to the last comment, when you say declaring the variable at the teams level would it look something like this  (In the code example I'm also thinking about how to implement this in RAJA teams): 

```
#pragma omp target ( this could be what the RAJA::launch method sets up) 
{

  #pragma omp teams distribute (RAJA team loop)
   for(e=0; e<ne_teams; ++e) { 

     double  sm_mem [ ]  // memory exclusive to each team

    #pragma parallel for (RAJA thread loop)
    for(t=0; t<threads; ++t) {
      sm_mem[t] = data; //sm_mem is available to all threads within a team. 
    }

  }
}
```

trws replied:
Yup, that's basically it.  Depending on the pattern it may be problematic to have it inside the distribute but that should *work*.  It likely will not perform well, compilers generally don't make that into shared memory (at least not yet), but it should work.

The form that should make it work *well* for this kind of case ought to look about like this:

```c
#pragma omp target ( this could be what the RAJA::launch method sets up) 
{

  #pragma omp teams distribute (RAJA team loop)
   for(e=0; e<ne_teams; ++e) { 

     double  sm_mem [ ]  // memory exclusive to each team
     #pragma omp allocate(omp_cgroup_mem_alloc:sm_mem)
    #pragma parallel for (RAJA thread loop)
    for(t=0; t<threads; ++t) {
      sm_mem[t] = t+1.5; //sm_mem is available to all threads within a team. 
    }

  }
}
```

The concern some people have right now is whether that's guaranteed to work the way we intended to by the text in the spec.  It's literally on the docket for discussion in the F2F next week.  I would think it should work, but whether implementers agree with me is an open question.

artv3 replied:
Cool! I think adding an omp target variant in would be great. I am partial to the version which does not require the `omp allocate` directive. The downside of the allocate directive is in the case where were have many arrays we would like to store in shared memory,  it can then get kind of verbose.  

How would you guys feel about adding the OMP target variant after we add the support in teams? As mentioned it would be useful for engaging with the compiler folks and illustrate patterns/styles we use.

trws replied:
It would be worth doing.  I'd recommend using a macro or something so that the allocate directive can be included or not though, I strongly suspect that the teams level declaration without it will become a big global memory allocation and perform *very* badly.  If that's something we want compilers to support better than that, it's a discussion we can have with them, but being able to compare the two would probably help a lot.

MrBurmark commented on src/common/HipDataUtils.hpp:
```diff
@@ -10,73 +10,69 @@
 /// Methods for HIP kernel data allocation, initialization, and deallocation.
 ///
 
```
What's up with all the formatting changes in this file?

artv3 replied:
Ah I must of ran clang-format on the files I modified.... We can revert this

artv3 replied:
Actually what do you guys think of running clang-format, in general?

rhornung67 replied:
In my experience, clang-format really screws up the formatting for complex template expressions making them much less readable. That's why I made a script for RAJA (in the scripts directory) that filters out the examples, exercises, etc. Here, the kernels rarely change after they are added. So I would prefer that the author of a kernel formats it in a way that makes it most readable. Then, don't change the formatting after that.

trws replied:
If possible, it would be nice to get a config such that we can use clang-format more widely than we do (can't say whether the specific issue is one with a fix yet, but if it is that would be nice).  Otherwise, can we start getting things that shouldn't be formatted by it marked appropriately with `//nofmt` comments so using it on the file as a whole doesn't mess things up?  It's a more useful tool if we don't have to think about it and can just always use it than if it has to be applied only in certain cases.

rhornung67 replied:
I agree for the most part. But sprinkling special comments all over the code is also a pain. In Axom we use clang-format only with clang10 and later. There was a substantial expansion of configuration options with the 10 release of the clang toolchain. I haven't had a chance to look into that with RAJA code. It may be workable, needing only a small number of format filtering comments in the code.

trws replied:
Yeah, the comments can be a pain (especially if there have to be a lot of them).  It's something that really only works if they can be used sparingly.

Setting a minimum clang-format version (higher than the required compiler version) seems completely reasonable to me.  It's best to set that to something anyway so people get consistent results.  Maybe something we can try and see how it goes.

rhornung67 replied:
If we decide to use clang-format as part of our regular development process, we should add a check for it in our azure CI.

trws replied:
Actually, this one looks like it had clang-format run on it *without the RAJA config for clang-format*. That's probably a non-trivial part of the difference.  Either that or some other formatter.  We have an explicit option set to avoid argument bin-packing and some other stuff that's happening here.

trws replied:
Also, we should use clang-format 12.  They finally added `IndentPragmas`!

rhornung67 replied:
I'll work on updating the clang-format config file to include more options it supports that we find useful in other projects. I'll put up a PR and we can debate it.

artv3 commented:
With the exception of the omp target variant, I think I addressed all the comments. 
Should we keep the omp target .cpp file or remove it? If its okay I was thinking we could add that variant once we  have support in Teams. Shouldn't be too difficult based on the discussions here.

artv3 commented on src/basic/MAT_MAT_SHARED.hpp:
```diff
@@ -0,0 +1,170 @@
+
+//Clang has trouble with the RAJA_TEAM_SHARED macros believing
+//its always compiling for the device
+#define MAT_MAT_SHARED_BODY_0_CLANG_HIP_CPU                   \
```
An additional detail, there seems to be confusion when compiling with clang + HIP. The compiler kept trying to invoke `__syncthreads()` in a CPU region even though we guard based on what we are compiling for.

MrBurmark replied:
When doing the device compile pass hipcc/clang will put in the device versions of the macros everywhere, in device functions, host device functions, and host only functions. Then it will make sure that code is valid everywhere, that's fine for device and host device functions, but it is not ok for host only functions. Nvcc doesn't look at host only code when it does the device pass so it doesn't see these kind of problems.

artv3 replied:
I'm going to pull your comment to the .hpp file :)

trws replied:
Crap, that's a good point.  That means we're going to have similar (though worse) issues with nvc++ compiling cuda.  We may want to think about how to re-work our macro guards to avoid the differences.

artv3 commented on src/basic/MAT_MAT_SHARED.hpp:
```diff
@@ -0,0 +1,170 @@
+  Real_ptr C = m_C;
+
+//Clang has trouble with the RAJA_TEAM_SHARED macros believing
+//its always compiling for the device
```
```suggestion
/*
 When doing the device compile pass hipcc/clang will put in the device
 versions of the macros everywhere, in device functions, host device functions, 
 and host only functions. Then it will make sure that code is valid everywhere, 
 that's fine for device and host device functions, but it is not ok for host only 
 functions. Nvcc doesn't look at host only code when it does the device pass 
 so it doesn't see these kind of problems.
 */
```

artv3 commented on src/basic/MAT_MAT_SHARED.hpp:
```diff
@@ -0,0 +1,170 @@
+  Real_ptr B = m_B;                                                            \
+  Real_ptr C = m_C;
+
+//Clang has trouble with the RAJA_TEAM_SHARED macros believing
```
```suggestion
```

rhornung67 commented on src/basic/MAT_MAT_SHARED.hpp:
```diff
@@ -0,0 +1,176 @@
+  if (Row < N && Col < N)                                                      \
+    C[Col + N * Row] = Cs[ty][tx];
+
+using launch_policy = RAJA::expt::LaunchPolicy<RAJA::expt::seq_launch_t
```
Something to think about: is it better/clearer to put the policy definitions in the header file or in the source files for the various back-end variants?  All the other kernels have them in the source files. I see why you put them here. But, it may be better to put them in the source files. That would avoid the need for macro guards in the launch policies, which can make them harder to read.

artv3 replied:
@rhornung67 , I moved some of the policies to the source files; however for the CPU based back-ends I still needed to add macro guards; this is on account of  the possibility of compiling with or without GPU support. Something @MrBurmark and I have chatted about is offering a way to specify a single policy even when compiling with the GPU -- this could clean up adding kernels to the perfsuite.

rhornung67 commented on src/common/RAJAPerfSuite.cpp:
```diff
@@ -20,6 +20,7 @@
 #include "basic/INIT_VIEW1D.hpp"
 #include "basic/INIT_VIEW1D_OFFSET.hpp"
 #include "basic/MULADDSUB.hpp"
+#include "basic/MAT_MAT_SHARED.hpp"
```
Please move this line and the ones below before the MULADDSUB line. They are in alphabetical order based on kernel name to make it easier to see that everything is defined consistently.

artv3 replied:
Fixed!

artv3 commented on src/common/RAJAPerfSuite.cpp:
```diff
@@ -403,6 +405,10 @@ KernelBase* getKernelObject(KernelID kid,
     }
+    case Basic_MAT_MAT_SHARED : {
+       kernel = new basic::MAT_MAT_SHARED(run_params);
+       break;
```
Ooops missed this one, working on a fix!

artv3 commented:
Thanks for the feedback folks, I think I've addressed all the comments.

artv3 commented:
Uh oh... we hit our pull rate limit: 

```
anyrequests: You have reached your pull rate limit. You may increase the limit by authenticating and upgrading: https://www.docker.com/increase-rate-limit.

See 'docker run --help'.

The command "docker run -u "$UID":"$GID" --rm -v ${TRAVIS_BUILD_DIR}/travis-data:/opt/intel/licenses -v ${TRAVIS_BUILD_DIR}:/home/raja rajaorg/compiler:$IMG ls" exited with 125.
```

rhornung67 commented:
I was afraid we would run into that. Travis is clamping down on their new limited support model. @mdavis36 is working on setting up azure CI for the perf suite based on what he did for RAJA. Hopefully, we can switch over to that soon.  Try restarting your travis builds in a couple of hours and see if it goes through.

artv3 commented:
@rhornung67 @MrBurmark , I just caught this branch up with develop. Can you please review?

artv3 commented:
Wait.. something is wrong, I have too many changes.

artv3 commented:
Please review: https://github.com/LLNL/RAJAPerf/pull/164
